File: runtime/pos/cuda_impl/patcher/cxx/to_cxx.cc
#include "to_cxx.h"

std::unique_ptr<std::vector<uint8_t>> to_cxx_vec(rust::Slice<const uint8_t> vec) {
    return std::make_unique<std::vector<uint8_t>>(
        vec.data(), vec.data() + vec.size()
    );
}

std::unique_ptr<std::string> to_cxx_string(rust::Str s) {
    return std::make_unique<std::string>(s.data(), s.size());
}

File: runtime/pos/cuda_impl/patcher/src/to_cxx.rs
#[cxx::bridge]
pub mod ffi {
    unsafe extern "C++" {
        include!("to_cxx.h");

        fn to_cxx_string(s: &str) -> UniquePtr<CxxString>;
        fn to_cxx_vec(v: &[u8]) -> UniquePtr<CxxVector<u8>>;
    }
}
